### ISSUE TITLE
☂️ Type-aware linter

### ISSUE THREAD

**Issue by Conaclos** [biomejs/biome#3187]
### Description

This umbrella issue tracks the development of type-aware lint rules.
We first motivate our decision to implement our own type synthesizer, and then present the type-aware rules we intend to implement. In a second section, we present preliminary design material that indicates the high-level direction we want to take.

## Motivation

Multiple [rules from TypeScript ESLint](https://typescript-eslint.io/rules/?=typeInformation) requires type information. Moreover, several linter rules could be enhanced by using type information.

TypeScript ESLint uses the TypeScript Compiler API to get types. This architecture has the advantage of using the TypeScript Compiler. However, it has several drawbacks:

- TSC is slow, and using the Compiler API makes the slowness even more noticeable.
  In fact, it is so slow that TypeScript ESLint provides presets with and without the rules that require type information.
- Biome and TSC use their own AST, which makes interoperability difficult.

This is why we think it is important to implement our own type synthesiser. If we have a fast type synthesiser, then we could enhance many lint rules with a marginal performance overhead. Note that we are not trying to implement a full-fledged type system or a type checker like TypeScript. Many attempt, even [the most promising](https://github.com/dudykr/stc) failed.

We believe that the Biome type synthesiser doesn't need to be perfect or handle complex TypeScript types. Even a rudimentary type synthesiser could be valuable to many lint rules.

## Type-aware lint rules

A first design and implementation of the Biome type synthesiser aims to implement a preliminary version of the following rules:

- useAwaitThenable ([await-thenable](https://typescript-eslint.io/rules/await-thenable/))

  Ensure that only thenable values are awaited. We could first target a rule that ensures that an awaited expression is a Promise.  We could ignore values with an unknown type.

- noFloatingPromises ([no-floating-promises](https://typescript-eslint.io/rules/no-floating-promises/))

  Ensure that a promise is handled (returned, awaited, ...).

- noForInArray ([no-for-in-array](https://typescript-eslint.io/rules/no-for-in-array/))

  Ensure that `for-in` is not used on arrays.

- noDuplicateLiteralEnumMembers ([no-duplicate-enum-values](https://typescript-eslint.io/rules/no-duplicate-enum-values))

  Ensure that every enum member initialized with a literal expression is unique. This doesn't necessarlly requires a type system. We need to compute literal expressions.

## Funding

To support this effort, please consider sponsoring Biome within our [Open Collective](https://opencollective.com/biome) or [GitHub sponsorship](https://github.com/sponsors/biomejs).

For companies wishing to speed up the development of this task, please consider hiring one of our contributors through our [Enterprise Support](https://biomejs.dev/enterprise/).

**Comment by MrTact** [comment]
Maybe the approach is to identify the most critical rules, and implement type inference __just for that type__? E.g. the first two rules appear to be promise-related, so figure out that a given identifier is a promise. This should still be built on a forward-looking type analyzer architecture, but by constraining the scope, you can make the effort much more manageable, since you don't have to achieve parity with tsc (a fool's errand anyway, which you point out). This also means that, post an admittedly non-trivial chunk of groundwork, additional rules can be developed (and shipped, ideally!) incrementally.

If the duplicate enum rule doesn't require a type checker, it should probably be peeled off into its own issue. Seems like it's bundled here just because it comes from typescript-eslint.

**Comment by henrik242** [comment]
I use `tsc --noEmit && biome check .` in my package.json. Can I drop tsc? What will I miss if I drop tsc?

**Comment by Conaclos** [comment]
> I use tsc --noEmit && biome check . in my package.json. Can I drop tsc? What will I miss if I drop tsc?

Biome doesn't perform type checking and is not intended to do in the future.
Once this issue is resolved, Biome will be able to implement some rules from Typescript ESLint plugin that requires type information.

**Comment by scamden** [comment]
Is it possible to give an extremely rough time frame for when this could become available? (Like is it on the order of weeks, months, or years).  

I ask because if it's weeks or even a few months I could probably wait for `noFloatingPromises` but if it's longer than 2-3 months I'll probably stick to eslint since that rules catches many many bugs. (Ran into two today which is why I'm here.)

**Comment by krzkaczor** [comment]
> I ask because if it's weeks or even a few months I could probably wait for noFloatingPromises but if it's longer than 2-3 months I'll probably stick to eslint since that rules catches many many bugs. (Ran into two today which is why I'm here.)

@scamden this is offtopic but in this [repo](https://github.com/marsfoundation/spark-app/blob/main/eslint.config.js) we run both biome (because it's 100x faster than eslint) and eslint (because of type-aware rules like `no-floating-promises`). This approach works fine for us. Eslint is faster than it used to be as well, because it only lints with a couple of rules, and we run it in parallel with tests anyway because it doesn't do any autofixing.

**Comment by scamden** [comment]
> > I ask because if it's weeks or even a few months I could probably wait for noFloatingPromises but if it's longer than 2-3 months I'll probably stick to eslint since that rules catches many many bugs. (Ran into two today which is why I'm here.)
> 
> @scamden this is offtopic but in this [repo](https://github.com/marsfoundation/spark-app/blob/main/eslint.config.js) we run both biome (because it's 100x faster than eslint) and eslint (because of type-aware rules like `no-floating-promises`). This approach works fine for us. Eslint is faster than it used to be as well, because it only lints with a couple of rules, and we run it in parallel with tests anyway because it doesn't do any autofixing.

Thanks for the idea @krzkaczor. Out of curiosity how does biome add value in that setup? Just trying to slightly reduce the runtime of the eslint by running fewer rules?

**Comment by krzkaczor** [comment]
@scamden yeah, and the ability to run eslint with tests in parallel (b/c eslint won't change any code). Overall it was more than 50% speedup in our setup.

**Comment by trungutt** [comment]
@scamden The link to the repo isn't available anymore. Could you please put it back? TIA

**Comment by junaga** [comment]
what about just shipping a default `tsconfig.json`, and then **tell**|**ask**|**suggest** Biome users in the docs to use it?

**Comment by testerez** [comment]
> Note that we are not trying to implement a full-fledged type system or a type checker like TypeScript

Does it mean that types will be properly identified only when they are explicit? Or will Biome be able to follow complex type inferences?

**Comment by ericchase** [comment]
wanted to chime in about
- noFloatingPromises ([no-floating-promises](https://typescript-eslint.io/rules/no-floating-promises/))

i think this is an extremely important rule. just this past week, i've forgotten to await async functions dozens of times. i had to pull in eslint with just 2 rules related to this to check my code. tsc is so horrendously slow.

i'm not really sure why these few specific rules need type awareness while other rules don't, but i would love to see these rules added!

**Comment by awhitford** [comment]
I'd like to share that [Astro Starlight](https://github.com/withastro/starlight/pull/1649#issuecomment-2010686715) cited the lack of a [no-floating-promises](https://typescript-eslint.io/rules/no-floating-promises/) equivalent was a show-stopper for adopting Biome over ESLint. 😞

**Comment by jpike88** [comment]
Special request for https://eslint.org/docs/latest/rules/require-await, I really miss that one and it's all related to ensuring that asynchronous code is written cleanly

**Comment by arendjr** [comment]
@jpike88 We should be able to implement that one already, since it’s syntax only. I’ve created a task for it: https://github.com/biomejs/biome/issues/4747

**Comment by 0x80** [comment]
> I should have checked first. We already have it: [biomejs.dev/linter/rules/use-await](https://biomejs.dev/linter/rules/use-await/) 😅

I feel confused 😅  I was watching this topic waiting for the equivalent of `@typescript-eslint/no-floating-promises` to arrive before making the jump to Biome, but now you're saying that it existed all along and doesn't even need type information? 🤔 

That's great news, if I didn't misunderstand it somehow...

The other ones I'd be interested in are `@typescript-eslint/no-misused-promises` and `@typescript-eslint/await-thenable` but those are not essential to me.

**Comment by dyc3** [comment]
No, we have the equivalent of `require-await`, not `no-floating-promises`. `useAwait` only enforces that an `await` appears in an async function.

**Comment by misl-smlz** [comment]
Are there any plans to support `no-floating-promises`?

**Comment by 0x80** [comment]
@misl-smlz See the first post in this issue?

**Comment by karlkeefer** [comment]
I just became a supporter of the project with this issue in mind. This would be a great addition to biome!

**Comment by gemue-parndt** [comment]
We really need no-floating-promises. This is such a big point on many projects but biome just doesn't have it...

**Comment by dyc3** [comment]
@kaykdm PRs are certainly welcome. For the multi file analysis, we are already discussing that in #3307, so I would hold off on that part for now.  It would be a great way to explore what type checking in biome could look like. You should be able to use our semantic model to track variable reassignments and such.

**Comment by arendjr** [comment]
I’m currently working on implementing multi-file analysis, which I think is an important prerequisite for getting type-enabled rules to work. As part of that, we’re also introducing data structures that track information from files that have been scanned in a project. In the `next` branch, we already have a first version of the `ProjectLayout`, which is important to be able to support monorepos. Right now I’m working on adding a `DependencyGraph` so that we can see which files import which others. The next step will be utilizing that to resolve imports to specific symbols inside files, and when we have that we can extend it with type information.

I don’t mind if you want to explore other parts of type inference already, but we’ll need to sync up at some point regarding the data model used, since it needs to be compatible with our multi-file approach. An important part of that is that we don’t want to lazily resolve imports and parse the files they point to, because such an approach makes it hard to cache and invalidate results. If you want you can look at how we construct and use the current `ProjectLayout`, because it should be somewhat similar to that, with entries keyed by path, and each values containing the symbols with their types for that file.

Let me know if there’s more you wish to discuss!

**Comment by kaykdm** [comment]
@dyc3 @arendjr 
Thank you for the reply and for letting me know about the ongoing discussion on multi-file analysis. I’ll focus solely on the type-checking aspect and leave the multi-file analysis for now. My first PR will target handling simple async functions.

**Comment by arendjr** [comment]
That would be a great start, thank you!

**Comment by ematipico** [comment]
I know that also @togami2864 is working on a prototype for the type checker. Maybe we could open a task or something for coordination

**Comment by mehulkar** [comment]
We recently [dropped eslint entirely for biome](https://x.com/mehulkar/status/1838807815462490480?s=46) at Vercel, but are considering re-introducing eslint due to the `noFloatingPromises` rule.

**Comment by kaykdm** [comment]
FYI: I created a task issue to track the remaining work. Please let me know if I'm missing anything.
https://github.com/biomejs/biome/issues/4956

**Comment by GermanJablo** [comment]
> This is why we think it is important to implement our own type synthesiser. If we have a fast type synthesiser, then we could enhance many lint rules with a marginal performance overhead. Note that we are not trying to implement a full-fledged type system or a type checker like TypeScript. Many attempt, even [the most promising](https://github.com/dudykr/stc) failed.
> 
> We believe that the Biome type synthesiser doesn't need to be perfect or handle complex TypeScript types. Even a rudimentary type synthesiser could be valuable to many lint rules.

1. Are you sure about this? I hope I'm wrong, but I have a hard time imagining how a type synthesizer that isn't fully tsc compliant could reliably warn about type-based rules.
2. Is the long-term goal to cover 100% of `typescript-eslint` rules, or just a subset?

**Comment by arendjr** [comment]
> Are you sure about this? I hope I'm wrong, but I have a hard time imagining how a type synthesizer that isn't fully `tsc` compliant could reliably warn about type-based rules.

The proof will be in the pudding, of course, but yes, I too believe we will be able to achieve (significant) value even where our type inference is still incomplete. The reason for this is that most use cases only require a subset of TypeScript features in order to determine whether something is of a given type. As long as we reliably fall back to an internal "unknown" type (for which we probably can just use TypeScript's actual `unknown` type) in the presence of unsupported type features, we can avoid our lint rules from triggering false positives (diagnostics for code that shouldn't trigger them) while allowing some false negatives (code that should trigger a diagnostic but doesn't) to remain.

This puts a lot of weight on the word "reliably" in your question. Our solution wouldn't be 100% reliable as long as we're not fully `tsc` compliant, because false negatives would still exist. But as long as we avoid the false positives (which are really obnoxious in lint rules) even 80% reliability can already bring a lot of value to users. Of course which percentage of reliability you will actually get depends a lot on what your code looks like and which libraries you use, so it may very well be a "your mileage may very" kinda situation. But the bet we're making is that we can offer significant value to many users, and we can probably improve our reliability over time.

> Is the long-term goal to cover 100% of typescript-eslint rules, or just a subset?

I don't think we even support of 100% of regular ESLint rules, so I don't see that as an explicit goal. But I do think that all the `typescript-eslint` rules should be theoretically implementable by us. It'll just be a matter of priority and complexity which ones we'll tackle and when.

**Comment by GermanJablo** [comment]
> Our solution wouldn't be 100% reliable as long as we're not fully tsc compliant, because false negatives would still exist. But as long as we avoid the false positives (which are really obnoxious in lint rules) even 80% reliability can already bring a lot of value to users


Thank you for the detailed response!

Personally, I'll stick with 100% typescript reliability, even if it comes at the cost of a slightly slower linter. That said, I don’t mean to undervalue the thought process behind your decision. I truly appreciate the tremendous work you’re doing and fully intend to use Biome as a formatter. But I guess as a linter it's not for me.

**Comment by togami2864** [comment]
I created a small type analyzer for linting. 
Of course, it's quite far from being practical, but it does similar things to typescript-eslint. If you are interested in the mechanism, please check it out 👀 
https://github.com/togami2864/type_poc_v3

**Comment by jpike88** [comment]
https://devblogs.microsoft.com/typescript/typescript-native-port/

Just leaving this here... I think our slow type checking days will soon be over anyway

**Comment by arendjr** [comment]
I understand Microsoft’s announcement for the native TypeScript compiler has people wondering how that will affect Biome’s plans for type inference. There’s a few things to consider here:

- Biome is a batteries-included tool where everything works out of the box. This is one of the reasons why we want to have our own type inference. It would work for users that work with plain JavaScript, or any situation where an external TypeScript compiler is not available.
- The speed argument against `tsc` will surely get weaker with the native compiler, but at the same time, an external service that we need to communicate with through IPC and which needs to retraverse and reparse all the files that Biome has already traversed and parsed would still be quite far from ideal.
- As the original post here mentioned: _because_ TypeScript parses everything on its own, we will have incompatible ASTs that we need to consolidate. It can be done, but it’s also a lot of work, and possibly not without its own issues.

That said, we were always open to the idea of interfacing with a TypeScript API, the same way typescript-eslint does. With the native compiler on the horizon, that option will likely only become more relevant. But that doesn’t exclude the possibility of implementing our own inference, which will likely be even faster and may work for more of our users.

And for those that do want to opt-in to retrieving type information from the TypeScript compiler, we are certainly not opposed to that option. It’s currently not on our roadmap, but if the new API from TypeScript is available at the end of the year, as they suggest, then who knows, it might be on next year’s :)

---

@GermanJablo Please consider that this thread is only about type **inference** and the lint rules we can build on top of that. Our intention was always that for type **checking** you’d keep on using `tsc`.

**Comment by DerTimonius** [comment]
not sure if it already has been mentioned before, but [`switch-exhaustiveness-check`](https://typescript-eslint.io/rules/switch-exhaustiveness-check) also needs type inference and would be a good addition.
do I mention it because having the rule missing after switching to Biome almost proved to be foot-gun? maybe 🙃

**Comment by lcswillems** [comment]
Would there be a way to leverage typescript-go and basically have all the TypeScript lint rules supported?

**Comment by ematipico** [comment]
@lcswillems 

There's already a comment that answers your question https://github.com/biomejs/biome/issues/3187#issuecomment-2716931205

**Comment by mdevils** [comment]
@arendjr do you think it's realistic to do inference without copying the whole TypeScript language service and keeping it up to date? `noFloatingPromises` is a good example: without knowing the type, you can't say it's a promise and not just a random class with `then` method.

**Comment by stunaz** [comment]
@mdevils he gave an answer here https://next.biomejs.dev/blog/vercel-partners-biome-type-inference/#how-can-you-write-a-type-informed-lint-rule-if-you-dont-know-the-type-information-is-correct

**Comment by arendjr** [comment]
Closing this issue as all the infrastructure is in place, and the first rules are already implemented. This will be available in Biome 2 which we intend to release Very Soon (TM).

**Comment by yairEO** [comment]
Consider changing the docs page for this rule, which currently shows this:

https://biomejs.dev/linter/rules/no-floating-promises/

![Image](https://github.com/user-attachments/assets/5aa8e34e-65c7-48b1-9af2-7b807ce48f37)

If there is currently no progress or planned tasks then this information might be outdated

**Comment by jpike88** [comment]
don't mean to necro an old issue but I see that 'useAwaitThenable' is in the above list of planned items. I don't find if it actually got implemented (https://github.com/biomejs/biome/discussions/7023)?

It appears to be a relatively harmless one, just an FYI I guess.

**Comment by Yuripetusko** [comment]
> don't mean to necro an old issue but I see that 'useAwaitThenable' is in the above list of planned items. I don't find if it actually got implemented ([#7023](https://github.com/biomejs/biome/discussions/7023))?
> 
> It appears to be a relatively harmless one, just an FYI I guess.

that's the only thing we're waiting for before making a switch. I used Biome on my personal projects and loved it, but at our current company we are very particular about some lint rules, and this one is a must. Love biome and can't wait for it to be implemented. Thanks devs and contributors

**Comment by ftzi** [comment]
Also just came here looking for `await-thenable`

**Comment by arendjr** [comment]
A PR for a first iteration of `useAwaitThenable` can be found here: https://github.com/biomejs/biome/pull/8341
